content/docs/user-guide/assets/asset-processor/debugging.md

```diff
@@ -6,9 +6,259 @@ weight: 500
+
+#### Solutions
+The primary solution to this problem is to update the builder to not cache information in a way that it persists across jobs.
+
+Automated tests can help prevent this from cropping up. You can see a test that covers the example case above [here](https://github.com/o3de/o3de/blob/development/AutomatedTesting/Gem/PythonTests/assetpipeline/fbx_tests/pythonassetbuildertests.py#L50). This test is annotated with how it's setup and why.
+
+#### Debugging
+This issue is difficult to identify because it requires a very specific setup, as explained previously. This issue was found by working backward from intermittent issues on automated asset processing jobs on the O3DE Jenkins servers.
+
+If your team is seeing situations where some people encounter different asset processing results than other team members, and you've already ruled out [a builder change without a version change](#NoVersionChange) as the source of the issue, then keep this in mind. Examine asset processing logs for the assets showing the issue, and check the asset processing order for cases where the issue occurs, compared against cases it does not occur. Specifically look for what other jobs ran on the same Asset Builder Executable, using the same Asset Builder on that executable.
+
+Asset Processor, both the command line and graphical interfaces, support a command line flag that can help with investigating. The flag "--sortJobsByDBSourceName" will stabilize the order that jobs are run in. Using this while debugging will let you test different job processing orders by renaming assets to control the order they run. The Asset Processor also allows for regset values to be controlled via command line, using the --regset flag with the setting to set. Specifically, setting the regset value /Amazon/AssetProcessor/Settings/Jobs/maxJobs to 1 will restrict the Asset Processor to only launch a single Asset Builder Executable. Note that if you have many assets to process, this will result in a long asset processing time, so it's recommended that you set this after you've processed all jobs in a previous Asset Processor run with more max jobs. Once you are running Asset Processor with one max job, all assets will process in that session on the same builder, letting you specifically process assets in the order of your choosing.
+
+Attaching Visual Studio to a running Asset Builder can also help with debugging. Following [these instruction](#DebugAssetBuilders) will run Asset Processor with a single builder, so you can process multiple assets in order with Visual Studio connected to the Asset Builder Executable. At that point you can try manually re-processing assets using the right click menu of the Asset Tab, and track what data is persisting across multiple jobs on the same Asset Builder on the same Asset Builder Executable.
```

Comment:
```suggestion
## Logic in Asset Builder persists between building assets

In some cases, due to information persisting during asset processing, you may get different Product Asset output from jobs based on the order Source Assets are processed. This is usually due to a bug within the Asset Builder that processed this content.

The Asset Processor starts one or more Asset Builder executables based on configuration settings, and uses these to process assets, sending jobs to these Asset Builder executables to run with the individual Asset Builders. There is no system within the Asset Builder Executable to completely shut down and clear all Asset Builders, so it is possible for an Asset Builder to persist information across multiple sessions of processing jobs for a single Asset Builder Executable.

### Effect

Stale data or logic in the Asset Builder Executable can result in an Asset processing incorrectly.

### Cause

The cause for this issue is when information is cached during one step in processing to make it easier to access at a later step.

### Example

This example covers a situation where this problem was encountered:

* The Scene Builder processes scene files, such as FBX files, into a collection of product assets.
* Each scene file has a scene manifest that contains additional processing rules.
* Within the scene manifest, a Python script file can be marked to run during scene processing.
* On initial startup, the Scene Builder within each Asset Builder Executable would start with no set Python script.
* On encountering a scene file that had a scene manifest with a Python script to run, the Scene Builder was storing this Python script to use later.
* The Scene Builder unintentionally stored this data in a way that persisted across jobs.
* For this example, FBX file `NoPython.fbx` does not have a Python script set to run in its Scene Manifest, but FBX `WithPython.fbx` does have a Python script set to run in its Scene Manifest.
* If both jobs end up running on the same Asset Builder Executable, then if `WithPython.fbx` is processed by the Scene Builder on that Asset Builder before `NoPython.fbx`, the Python script to run was persisted in a way that it was run on `NoPython.fbx`.
* The end result was `NoPython.fbx` was producing different output if it ran after `WithPython.fbx` on the same Asset Builder Executable as `WithPython.fbx` ran on over any other situation: Running before `WithPython.fbx`, running on a different Asset Builder Executable.

### Solutions

The primary solution to this problem is to update the builder to not cache information in a way that it persists across jobs.

Automated tests can help prevent this from cropping up. You can see a test that covers the example case above [here](https://github.com/o3de/o3de/blob/development/AutomatedTesting/Gem/PythonTests/assetpipeline/fbx_tests/pythonassetbuildertests.py#L50). This test is annotated with how it's setup and why.

### Debugging

This issue is difficult to identify because it requires a very specific setup, as explained previously. This issue was found by working backward from intermittent issues on automated asset processing jobs on the O3DE Jenkins servers.

If your team is seeing situations where some people encounter different asset processing results than other team members, and you've already ruled out [a builder change without a version change](#NoVersionChange) as the source of the issue, then keep this in mind. Examine asset processing logs for the assets showing the issue, and check the asset processing order for cases where the issue occurs, compared against cases it does not occur. Specifically look for what other jobs ran on the same Asset Builder Executable, using the same Asset Builder on that executable.

Asset Processor, both the command line and graphical interfaces, support a command line flag that can help with investigating. The flag `--sortJobsByDBSourceName` will stabilize the order that jobs are run in. Using this while debugging will let you test different job processing orders by renaming assets to control the order they run. The Asset Processor also allows for regset values to be controlled via command line, using the `--regset` flag with the setting to set. Specifically, setting the regset value `/Amazon/AssetProcessor/Settings/Jobs/maxJobs` to `1` will restrict the Asset Processor to only launch a single Asset Builder executable. Note that if you have many assets to process, this will result in a long asset processing time, so it's recommended that you set this after you've processed all jobs in a previous Asset Processor run with more max jobs. Once you are running Asset Processor with one max job, all assets will process in that session on the same builder, letting you specifically process assets in the order of your choosing.

Attaching Visual Studio to a running Asset Builder can also help with debugging. Following [these instructions](#DebugAssetBuilders) will run Asset Processor with a single builder, so you can process multiple assets in order with Visual Studio connected to the Asset Builder Executable. At that point you can try manually re-processing assets using the right click menu of the Asset Tab, and track what data is persisting across multiple jobs on the same Asset Builder on the same Asset Builder Executable.
```

I think the example provided here could be greatly simplified, but without the full technical details, I'm unsure how. It looks like a test artifact was being cached and inappropriately loaded for another test, depending on the order of the test runs?

Reply:
> I think the example provided here could be greatly simplified, but without the full technical details, I'm unsure how. It looks like a test artifact was being cached and inappropriately loaded for another test, depending on the order of the test runs?

It's a complex situation that I'm struggling to figure out how to describe in a simple way. The Asset Builder Executables run whatever code you give them. So if you're not careful as an Asset Builder author, you can accidentally use information from an asset processed earlier to change how an asset is processed later.

Reply:
Will take an issue that this example could either be simplified (to be in-line) or expanded (as a separate example of conflicts, with appropriate resources available.)